File: src/main/java/org/apache/prepbuddy/groupingops/LevenshteinDistance.java
package org.apache.prepbuddy.groupingops;

import org.apache.commons.lang.StringUtils;
import scala.Tuple2;

import java.util.ArrayList;
import java.util.List;

public class LevenshteinDistance implements ClusteringAlgorithm {
    @Override
    public Clusters getClusters(List<Tuple2<String, Integer>> tuples) {
        Clusters clusters = new Clusters();

        List<Integer> indexes = new ArrayList<>();
        for (int i = 0; i < tuples.size(); i++) {
            Tuple2<String, Integer> tuple = tuples.get(i);
            String tupleKey = tuple._1();
            clusters.add(tupleKey, tuple);
            for (int j = i + 1; j < tuples.size(); j++) {
                Tuple2<String, Integer> otherTuple = tuples.get(j);
                String otherTupleKey = otherTuple._1();
                int distance = StringUtils.getLevenshteinDistance(tupleKey, otherTupleKey);
                if (distance < 4 && !(indexes.contains(j))) {
                    clusters.add(tupleKey, otherTuple);
                    indexes.add(j);
                }
            }
        }
        return clusters;
    }
}

File: src/main/scala/org/apache/datacommons/prepbuddy/clusterers/LevenshteinDistance.scala
package org.apache.datacommons.prepbuddy.clusterers

import org.apache.commons.lang.StringUtils

import scala.collection.mutable.ListBuffer

/**
  * This algorithm treats cardinal value as a key but grouped those values
  * together whom Lenenshtein Distance is less than 4.
  */
class LevenshteinDistance extends ClusteringAlgorithm {
    override def getClusters(tuples: Array[(String, Int)]): Clusters = {
        val clusters: Clusters = new Clusters
        val indexes: ListBuffer[Int] = ListBuffer.empty
        tuples.view.zipWithIndex.foreach {
            case (tuple, index) =>
                val key: String = tuple._1
                clusters.add(key, tuple)
                var secondIndex: Int = index + 1
                while (secondIndex < tuples.length) {
                    val anotherTuple = tuples(secondIndex)
                    val anotherKey: String = anotherTuple._1
                    val distance: Int = StringUtils.getLevenshteinDistance(key, anotherKey)
                    if (distance < 4 && !indexes.contains(secondIndex)) {
                        clusters.add(key, anotherTuple)
                        indexes += secondIndex
                    }
                    secondIndex += 1
                }
        }
        clusters
    }
}


